### PR TITLE
egsp stream validation rules

### PR DESCRIPTION
**Pull request by nahguam**
# stream-registry PR

egsp stream validation rules

# PR Checklist Forms

- [ ] CHANGELOG.md updated
- [ ] Reviewer assigned
- [ ] PR assigned (presumably to submitter)
- [ ] Labels added (enhancement, bug, documentation)

### CODE REVIEW

File: handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java [hunk 1]
```diff
@@ -15,31 +15,71 @@
...
 import com.expediagroup.streamplatform.streamregistry.model.Specification;
 import com.expediagroup.streamplatform.streamregistry.model.Stream;
 
 @Component
-public class EgspKafkaStreamHandler<E extends Stream> implements Handler<E> {
-
+public class EgspKafkaStreamHandler implements Handler<Stream> {
   @Override
   public String type() {
     return "egsp.kafka";
```
Comment (rshotels) [inline handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java]: Consider declaring a constant /  using one from an existing class (if there is one)

File: handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java [hunk 2]
```diff
@@ -15,31 +15,71 @@
...
+  }
+
+  private Specification handle(Stream stream) {
+    Specification specification = stream.getSpecification();
+    ObjectNode configuration = specification.getConfiguration();
+
+    ObjectNode log = (ObjectNode) configuration.get("log");
+    checkNotNull(log, "log must not be null");
+
+    validateGreaterThanZero(log, "partitions");
```
Comment (rshotels) [inline handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java]: Should we declare these field names as constants?

File: handler-egsp/src/test/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+import com.expediagroup.streamplatform.streamregistry.model.Specification;
+import com.expediagroup.streamplatform.streamregistry.model.Stream;
+
+public class EgspKafkaStreamHandlerTest {
+  private final ObjectMapper mapper = new ObjectMapper();
+  private final EgspKafkaStreamHandler underTest = new EgspKafkaStreamHandler();
+
+  @Test
+  public void type() {
+    assertThat(underTest.type(), is("egsp.kafka"));
```
Comment (rshotels) [inline handler-egsp/src/test/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandlerTest.java]: egsp.kafka should be a shared constant

File: handler-egsp/src/test/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  @Test
+  public void insertTypical() {
+    underTest.handleInsert(stream(6, 3, "delete", days(7), days(1)));
+  }
+
+  @Test(expected = NullPointerException.class)
+  public void insertNullPartitions() {
+    underTest.handleInsert(stream(null, 3, "delete", days(7), days(1)));
```
Comment (rshotels) [inline handler-egsp/src/test/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandlerTest.java]: Consider creating constants ONE_PARTITION MANY_PARTITIONS, ONE_DAY, MANY_DAYS, DELETE, etc.

File: handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java [hunk 5]
```diff
@@ -15,31 +15,71 @@
...
   @Override
-  public Specification handleUpdate(E stream, E existing) throws HandlerException {
-    return stream.getSpecification();
+  public Specification handleUpdate(Stream stream, Stream existing) {
+    return handle(stream);
+  }
+
+  private Specification handle(Stream stream) {
+    Specification specification = stream.getSpecification();
+    ObjectNode configuration = specification.getConfiguration();
```
Comment (rshotels) [inline handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java]: Can specification be null?

File: handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java [hunk 6]
```diff
@@ -15,31 +15,71 @@
...
+
+    ObjectNode log = (ObjectNode) configuration.get("log");
+    checkNotNull(log, "log must not be null");
+
+    validateGreaterThanZero(log, "partitions");
+    validateGreaterThanZero(log, "replicationFactor");
```
Comment (vedantchokshi) [inline handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java]: Can we change this to `replication.factor` to match all the other config naming conventions?
```suggestion
    validateGreaterThanZero(log, "replication.factor");
```

File: handler-egsp/src/test/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+      Long retentionMs,
+      Long minCompactionLagMs) {
+    ObjectNode config = mapper.createObjectNode();
+    ObjectNode log = config.putObject("log");
+
+    if (partitions != null) {
+      log.put("partitions", partitions);
+    }
+    if (replicationFactor != null) {
+      log.put("replicationFactor", replicationFactor);
```
Comment (vedantchokshi) [inline handler-egsp/src/test/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandlerTest.java]: Changing here too
```suggestion
      log.put("replication.factor", replicationFactor);
```

File: handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java [hunk 8]
```diff
@@ -15,31 +15,71 @@
...
+
+    Predicate<JsonNode> predicate = value -> List.of("delete", "compact").contains(value.asText());
+    validate(log, "cleanup.policy", predicate, "%s must be one of delete or compact");
+
+    if ("delete".equals(log.get("cleanup.policy").asText())) {
+      validateGreaterThanZero(log, "retention.ms");
```
Comment (vedantchokshi) [inline handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java]: Seeing as this in ms, should we add a lower bound besides 0 if the users enter ridiculously low numbers? I am not sure what the impact of that could be

File: handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java (hunk 1 above)
Comment (nahguam) [reply handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java]: done

File: handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java (hunk 2 above)
Comment (nahguam) [reply handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java]: where they're only used once or twice I'm fine with inlined.

File: handler-egsp/src/test/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandlerTest.java (hunk 3 above)
Comment (nahguam) [reply handler-egsp/src/test/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandlerTest.java]: done

File: handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java (hunk 5 above)
Comment (nahguam) [reply handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java]: no. graphql enforces that

File: handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java (hunk 6 above)
Comment (nahguam) [reply handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java]: done

File: handler-egsp/src/test/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandlerTest.java (hunk 7 above)
Comment (nahguam) [reply handler-egsp/src/test/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandlerTest.java]: done

File: handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java (hunk 8 above)
Comment (nahguam) [reply handler-egsp/src/main/java/com/expediagroup/streamplatform/streamregistry/handler/EgspKafkaStreamHandler.java]: The log cleaner still only kicks in periodically. So it wouldn't create any extra burden on the infrastructure. It only affects how much data in the log is not cleaned.

**Pull request merged**